torch/utils/_benchmark/utils/valgrind_wrapper.py

```diff
@@ -0,0 +1,227 @@
+        return fn_diff_counts
+
+    def _invoke(self, stmt: str, setup: str, number: int, num_threads: int):
+        working_dir = tempfile.mkdtemp()
+
+        _, script_file = tempfile.mkstemp(prefix="timer_callgrind_", suffix=".py", dir=working_dir)
+        _, callgrind_out = tempfile.mkstemp(prefix="callgrind.", suffix=".out", dir=working_dir)
+        _, error_log = tempfile.mkstemp(prefix="error_", suffix=".txt", dir=working_dir)
+
+        try:
+            with open(script_file, "wt") as f:
+                f.write(self._construct_script(
+                    stmt=stmt, setup=setup, number=number,
+                    num_threads=num_threads, error_log=error_log))
+
+            valgrind_invocation = subprocess.run(
+                [
+                    "valgrind",
+                    "--tool=callgrind",
+                    f"--callgrind-out-file={callgrind_out}",
+                    "--dump-line=yes",
+                    "--dump-instr=yes",
+                    "--collect-jumps=yes",
+                    "--instr-atstart=no",
+                    "python",
+                    script_file,
+                ],
+                capture_output=True
+            )
+            if valgrind_invocation.returncode:
+                with open(error_log, "rt") as f:
+                    error_report = f.read()
+                if not error_report:
+                    error_report = "Unknown error."
+                    error_report += "\n" + valgrind_invocation.stdout.decode("utf-8")
+                    error_report += "\n" + valgrind_invocation.stderr.decode("utf-8")
+                raise OSError(f"Failed to collect callgrind profile:\n{error_report}")
+
+            annotate_invocation = subprocess.run(
+                [
+                    "callgrind_annotate",
+                    "--inclusive=no",
+                    callgrind_out
+                ],
+                capture_output=True,
+                check=True,
+            )
+
+            begin_collecting = False
+            fn_counts = []
+            for l in annotate_invocation.stdout.decode("utf-8").splitlines(keepends=False):
+                if not begin_collecting and re.match(r"Ir\s+file:function", l):
+                    begin_collecting = True
+                    continue
+
+                count_match = re.match(r"^\s*([0-9,]+)\s+(.+:.+)$", l)
+                if count_match:
+                    ir, file_function = count_match.groups()
+                    ir = int(ir.replace(",", ""))
+                    fn_counts.append((ir, file_function))
+                    continue
+
+                if begin_collecting and re.match(r"-+", l):
+                    continue
+
+                begin_collecting = False
+
+            return fn_counts
+        finally:
+            shutil.rmtree(working_dir)
+
+    @staticmethod
+    def _construct_script(stmt: str, setup: str, number: int, num_threads: int, error_log: str):
+        return textwrap.dedent(r"""
```

Comment:
Another option, instead of having such a long amount  of generated code, is to just put this in a module in torch and then use it from here. But I would hope that we can reduce this code by doing Dmytro's suggestion of doing the magic assembly instructions; I like that a lot better.

Reply:
The magic assembly instructions don't shorten the generated code much, but they make it much clearer. I considered having a real python file with a line like `from generated_sections import setup_fn, stmt_fn` where things could hook in, but I think it's short enough and has enough stuff templated in that having everything in one place is easier. However if it's still too much generated code for folks' taste I'm happy to switch to the other pattern.